jax/_src/pallas/core.py

```diff
@@ -327,42 +335,51 @@ def _tile_ref(ref: state.AbstractRef, block_shape: tuple[int, ...] | None
   return ref.update(inner_aval=ref.inner_aval.update(shape=shape))
 
 
-def _check_static_ref_shape(ref: state.AbstractRef) -> state.AbstractRef:
-  shape = ref.shape
-  if not jax_core.is_constant_shape(shape):
-    # TODO(necula): thread the tree labels so that we can localize the error
-    raise ValueError("shape polymorphism for Pallas does not support "
-                     f"dynamically-shaped blocks. Found block_shape: {shape}")
-  return ref
+def _get_ref_avals(grid,
+                   in_avals: Sequence[jax_core.ShapedArray],
+                   in_specs: Sequence[BlockSpec],
+                   in_paths: Sequence[tree_util.KeyPath],
+                   out_avals: Sequence[jax_core.ShapedArray],
+                   out_specs: Sequence[BlockSpec],
+                   out_paths: Sequence[tree_util.KeyPath]):
+  def make_ref_aval(aval: jax_core.ShapedArray,
+                    spec: BlockSpec,
+                    path: tree_util.KeyPath,
+                    what: str) -> state.AbstractRef:
+    if spec is no_block_spec:
+      memory_space = None
+      block_shape = None
+    else:
+      memory_space = spec.memory_space
+      block_shape = spec.block_shape
 
+    ref_aval = AbstractMemoryRef(aval, memory_space)
+    if block_shape is not None:
+      if len(ref_aval.shape) != len(block_shape):
+        raise ValueError(
+            f"Block shape for {what}{tree_util.keystr(path)} (= {block_shape}) "
```

Comment:
Should we say block size must have same rank as array to be more clear? Since length is a little ambiguous for arrays.

Reply:
I changed it to "shape should have the same number of dimensions as the array shape". I did not use "rank" because that is not a property of the shape, but of the array, I think.